File: src/VerletListAdress.cpp
/*
  Copyright (C) 2012,2013,2014,2015,2016,2017(H)
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include "VerletListAdress.hpp"
#include "Real3D.hpp"
#include "Particle.hpp"
#include "Cell.hpp"
#include "System.hpp"
#include "storage/Storage.hpp"
#include "bc/BC.hpp"
#include "iterator/CellListAllPairsIterator.hpp"
#include "iterator/CellListIterator.hpp"

namespace espressopp {

  using namespace espressopp::iterator;

  LOG4ESPP_LOGGER(VerletListAdress::theLogger, "VerletList");

  /*-------------------------------------------------------------*/

    VerletListAdress::VerletListAdress(shared_ptr<System> system, real cut, real adrCut,
                                       bool rebuildVL, real _dEx, real _dHy)
    :SystemAccess(system) {
      LOG4ESPP_INFO(theLogger, "construct VerletList, cut = " << cut);

      if (!system->storage) {
         throw std::runtime_error("system has no storage");
      }
      skin = system->getSkin();
      cutverlet = cut + skin;
      cutsq = cutverlet * cutverlet;
      builds = 0;
      dEx = _dEx;
      dHy = _dHy;
      adrCenterSet = false;
      real adressSize = dEx + dHy + skin; // adress region size
      if (dEx + dHy == 0) adressSize = 0; // 0 should be 0
      adrsq = adressSize * adressSize;
      adrCutverlet = adrCut + skin;
      adrcutsq = adrCutverlet*adrCutverlet;

      if (rebuildVL) rebuild(); // not called if exclutions are provided

      // make a connection to System to invoke rebuild on resort
      connectionResort = system->storage->onParticlesChanged.connect(
          boost::bind(&VerletListAdress::rebuild, this));
    }

    /*-------------------------------------------------------------*/


    void VerletListAdress::rebuild()
    {
      vlPairs.clear();
      adrZone.clear(); // particles in adress zone
      cgZone.clear(); // particles in CG zone
      adrPairs.clear(); // pairs in adress zone
      const bc::BC& bc = *getSystemRef().bc;

      // get local cells
      CellList localcells = getSystem()->storage->getLocalCells();

      // if adrCenter is not set, the center of adress zone moves along with some particles
      if (!adrCenterSet) {
          // loop over all VP particles (reals and ghosts) on node
          for (CellListIterator it(localcells); it.isValid(); ++it) {

                // loop over positions
                for (std::vector<Real3D*>::iterator it2 = adrPositions.begin(); it2 != adrPositions.end(); ++it2){
                    Real3D dist;
                    real distsq;
                    bc.getMinimumImageVectorBox(dist, it->getPos(), **it2);

                    if (getAdrRegionType()){ // spherical adress region
                       distsq=dist.sqr();
                    }
                    else {  // slab-type adress region
                       distsq=dist[0]*dist[0];
                    }
                    if (distsq <= adrsq) {
                        adrZone.insert(&(*it));
                        break; // do not need to loop further
                    }
                }
                // if not near enough to any adrPositions, put in cgZone
                if (adrZone.count(&(*it)) == 0) {
                    cgZone.insert(&(*it));
                }
          }
      }
      // center of adress zone is fixed
      else {
          for (CellListIterator it(localcells); it.isValid(); ++it) {
              Real3D dist;
              real distsq;
              bc.getMinimumImageVectorBox(dist, it->getPos(), adrCenter);
              //Real3D dist = it->getPos() - adrCenter;                                   					        // CHANGE FOR X SPLIT VS SPHERE
              //real dist = it->getPos()[0] - adrCenter[0];                                  					        // CHANGE FOR X SPLIT VS SPHERE
              //real distsq = dist.sqr();                                                                                                 // CHANGE FOR X SPLIT VS SPHERE
              //real distsq = dist*dist;                                                                                                // CHANGE FOR X SPLIT VS SPHERE
              if (getAdrRegionType()){ // spherical adress region
                distsq=dist.sqr();
              }
              else {  // slab-type adress region
                distsq=dist[0]*dist[0];
              }
              if (distsq <= adrsq) {
                  adrZone.insert(&(*it));
              }
              else {
                  cgZone.insert(&(*it));
              }
          }

      }

      // add particles to adress pairs and VL
      CellList cl = getSystem()->storage->getRealCells();
      int count=0;

      for (CellListAllPairsIterator it(cl); it.isValid(); ++it) {
        checkPair(*it->first, *it->second);
        count+=1;
      }

      LOG4ESPP_INFO(theLogger, "rebuilt VerletList, cutsq = " << cutsq
                   << " local size = " << vlPairs.size());
      builds++;
    }

    /*-------------------------------------------------------------*/

    void VerletListAdress::checkPair(Particle& pt1, Particle& pt2)
    {

      Real3D d = pt1.position() - pt2.position();
      real distsq = d.sqr();

      LOG4ESPP_TRACE(theLogger, "p1: " << pt1.id()
                     << " @ " << pt1.position()
             << " - p2: " << pt2.id() << " @ " << pt2.position()
             << " -> distsq = " << distsq);

      // see if it's in the exclusion list (both directions, CG particles only)
      if (exList.count(std::make_pair(pt1.id(), pt2.id())) == 1) return;
      if (exList.count(std::make_pair(pt2.id(), pt1.id())) == 1) return;
      // see if it's in the adress zone
      if (adrZone.count(&pt1) == 1 || adrZone.count(&pt2) == 1) {
          if (distsq > adrcutsq) return;
          adrPairs.add(pt1, pt2); // add to adress pairs
      }
      else {
          if (distsq > cutsq) return;
          vlPairs.add(pt1, pt2); // add pair to Verlet List
      }
    }

    /*-------------------------------------------------------------*/

    int VerletListAdress::totalSize() const
    {
      System& system = getSystemRef();
      int size = vlPairs.size();
      int allsize;

      mpi::all_reduce(*system.comm, size, allsize, std::plus<int>());
      return allsize;
    }


    bool VerletListAdress::exclude(longint pid1, longint pid2) {
        exList.insert(std::make_pair(pid1, pid2));
        return true;
    }

    void VerletListAdress::addAdrParticle(longint pid) {
          adrList.insert(pid);
    }

    void VerletListAdress::setAdrCenter(real x, real y, real z){
        adrCenter = Real3D(x, y, z);
        adrCenterSet = true;
        adrPositions.push_back(&adrCenter);
    }

    Real3D VerletListAdress::getAdrCenter(){return adrCenter;}

    void VerletListAdress::setAdrRegionType(bool _sphereAdr){
        sphereAdr = _sphereAdr;
    }

    bool VerletListAdress::getAdrRegionType(){
        return sphereAdr;
    }

    /*-------------------------------------------------------------*/

    VerletListAdress::~VerletListAdress()
    {
      LOG4ESPP_INFO(theLogger, "~VerletList");

      if (!connectionResort.connected()) {
        connectionResort.disconnect();
      }
    }

    /****************************************************
    ** REGISTRATION WITH PYTHON
    ****************************************************/

    void VerletListAdress::registerPython() {
      using namespace espressopp::python;

      bool (VerletListAdress::*pyExclude)(longint pid1, longint pid2)
            = &VerletListAdress::exclude;

      void (VerletListAdress::*pyAddAdrParticle)(longint pid)
            = &VerletListAdress::addAdrParticle;

      void (VerletListAdress::*pySetAdrCenter)(real x, real y, real z)
                  = &VerletListAdress::setAdrCenter;

      void (VerletListAdress::*pySetAdrRegionType)(bool _sphereAdr)
            = &VerletListAdress::setAdrRegionType;

      class_<VerletListAdress, shared_ptr<VerletList> >
        ("VerletListAdress", init< shared_ptr<System>, real, real, bool, real, real>())
        .add_property("system", &SystemAccess::getSystem)
        .add_property("builds", &VerletListAdress::getBuilds, &VerletListAdress::setBuilds)
        .def("totalSize", &VerletListAdress::totalSize)
        .def("exclude", pyExclude)
        .def("addAdrParticle", pyAddAdrParticle)
        .def("setAdrCenter", pySetAdrCenter)
        .def("setAdrRegionType", pySetAdrRegionType)
        .def("rebuild", &VerletListAdress::rebuild)
        ;
    }

}

File: src/VerletListAdress.py
#  Copyright (C) 2012,2013,2015
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
***************************
espressopp.VerletListAdress
***************************

The VerletListAdress is the Verlet List to be used for AdResS or H-AdResS
simulations. When creating the VerletListAdress one has to provide the system
and specify both cutoff for the CG interaction and adrcutoff for the atomistic
interaction. Often, it is important to set the atomistic adrcutoff much bigger
than the actual interaction's cutoff would be, since also the atomistic part of
the VerletListAdress (adrPairs) is built based on the coarse-grained particle
positions. For a much larger coarse-grained cutoff it is for example possible
to also set the atomistic cutoff on the same value as the coarse-grained one.

Furthermore, the sizes of the explicit and hybrid region have to be
provided (dEx and dHy in the example below) and the center of the atomistic
region has to be set (adrCenter). Additionally, it can be chosen between a spherical and a slab-like geometry (sphereAdr).

The AdResS region can also be defined based on one or more particles. For a single particle, in this case a spherical region moves along with the particle. For many such region defining particles, the high-resolution/hybrid region corresponds to the overlap of the different spherical regions based on the individual particles (for details see Kreis et al., JCTC doi: 10.1021/acs.jctc.6b00440). Note that more region defining particles mean a higher computational overhead as these particles need to be communicated among all processors (also see explanations in AdResS.py). Also note that region defining particles should be normal/CG particles, not atomistic/AdResS ones.

**Bascially the VerListAdress provides 4 lists:**

* adrZone: A list which holds all particles in the atomistic and hybrid region
* cgZone: A list which holds all particles in the coarse-grained region
* adrPairs: A list which holds all pairs which have at least one particle in the
  adrZone, i.e. in the atomistic or hybrid region
* vlPairs: A list which holds all pairs which have both particles in the cgZone,
  i.e. in the coarse-grained region

Example - creating the VerletListAdress for a slab-type adress region fixed in space (only the x value of adrCenter is used):

>>> vl      = espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, adrCenter=[Lx/2, Ly/2, Lz/2])

or

>>> vl      = espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, adrCenter=[Lx/2, Ly/2, Lz/2], sphereAdr=False)

Example - creating the VerletListAdress for a spherical adress region centered on adrCenter and fixed in space:

>>> vl      = espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, adrCenter=[Lx/2, Ly/2, Lz/2], sphereAdr=True)

Example - creating the VerletListAdress for a spherical adress region centered on one particle and moving with the particle

>>> vl      = espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, pids=[adrCenterPID], sphereAdr=True)

Example - creating the VerletListAdress for a adress region based on the overlapping spherical regions by several particles

>>> vl      = espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, pids=[adrCenterPID1,adrCenterPID2,adrCenterPID3, ... ], sphereAdr=True)

.. function:: espressopp.VerletListAdress(system, cutoff, adrcut, dEx, dHy, adrCenter, pids, exclusionlist, sphereAdr)

		:param system:
		:param cutoff:
		:param adrcut:
		:param dEx:
		:param dHy:
		:param adrCenter: (default: [])
		:param pids: (default: [])
		:param exclusionlist: (default: [])
		:param sphereAdr: (default: False)
		:type system:
		:type cutoff:
		:type adrcut:
		:type dEx:
		:type dHy:
		:type adrCenter:
		:type pids:
		:type exclusionlist:
		:type sphereAdr:

.. function:: espressopp.VerletListAdress.addAdrParticles(pids, rebuild)

		:param pids:
		:param rebuild: (default: True)
		:type pids:
		:type rebuild:
		:rtype:

.. function:: espressopp.VerletListAdress.exclude(exclusionlist)

		:param exclusionlist:
		:type exclusionlist:
		:rtype:

.. function:: espressopp.VerletListAdress.rebuild()

		:rtype:

.. function:: espressopp.VerletListAdress.totalSize()

		:rtype:
"""

from espressopp import pmi
import _espressopp
import espressopp
from espressopp.esutil import cxxinit

class VerletListAdressLocal(_espressopp.VerletListAdress):


    def __init__(self, system, cutoff, adrcut, dEx, dHy, adrCenter=[], pids=[], exclusionlist=[], sphereAdr=False):

        if pmi.workerIsActive():
            cxxinit(self, _espressopp.VerletListAdress, system, cutoff, adrcut, False, dEx, dHy)
            #self.cxxclass.setAtType(self, atType)
            # check for exclusions
            if (exclusionlist != []):
                # add exclusions
                for pair in exclusionlist:
                    pid1, pid2 = pair
                    self.cxxclass.exclude(self, pid1, pid2)
            # add adress particles
            if (pids != []):
                for pid in pids:
                    self.cxxclass.addAdrParticle(self, pid)
            # set adress center
            if (adrCenter != []):
                self.cxxclass.setAdrCenter(self, adrCenter[0], adrCenter[1], adrCenter[2])
            # set adress region type (slab or spherical)
            self.cxxclass.setAdrRegionType(self,sphereAdr)

            # rebuild list now
            self.cxxclass.rebuild(self)


    def totalSize(self):

        if pmi.workerIsActive():
            return self.cxxclass.totalSize(self)


    def exclude(self, exclusionlist):
        """
        Each processor takes the broadcasted exclusion list
        and adds it to its list.
        """
        if pmi.workerIsActive():
            for pair in exclusionlist:
                pid1, pid2 = pair
                self.cxxclass.exclude(self, pid1, pid2)
            # rebuild list with exclusions
            self.cxxclass.rebuild(self)


    def addAdrParticles(self, pids, rebuild=True):
        """
        Each processor takes the broadcasted atomistic particles
        and adds it to its list.
        """
        if pmi.workerIsActive():
            for pid in pids:
                self.cxxclass.addAdrParticle(self, pid)
            if rebuild:
                # rebuild list with adress particles
                self.cxxclass.rebuild(self)

    def rebuild(self):
        if pmi.workerIsActive():
            self.cxxclass.rebuild(self)

if pmi.isController:
    class VerletListAdress(object):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls = 'espressopp.VerletListAdressLocal',
            pmiproperty = [ 'builds' ],
            pmicall = [ 'totalSize', 'exclude', 'addAdrParticles', 'rebuild' ]
            )
